{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "You can find the C code in this temporary file: C:\\Users\\Sean\\AppData\\Local\\Temp\\theano_compilation_error__xicqdsu\n"
     ]
    },
    {
     "ename": "Exception",
     "evalue": "(\"Compilation failed (return status=1): C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `run':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:99: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:124: undefined reference to `__imp_PyExc_ValueError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:130: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:158: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:195: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:210: undefined reference to `__imp_PyExc_ValueError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:479: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `_Py_INCREF':\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/include/object.h:459: undefined reference to `__imp__Py_NoneStruct'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `run':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:485: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:244: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:265: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:173: undefined reference to `__imp_PyExc_TypeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:179: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:515: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:514: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:516: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:259: undefined reference to `__imp_PyExc_TypeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:279: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:294: undefined reference to `__imp_PyExc_ValueError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:328: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:349: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:216: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:343: undefined reference to `__imp_PyExc_TypeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:300: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:397: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `instantiate':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:550: undefined reference to `__imp_PyExc_TypeError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `_import_array':\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1493: undefined reference to `__imp_PyCapsule_Type'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1494: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `PyInit_mf6917bb35eaa79d4a20c20b9dc13c0435e656b1bdb67265fed3d06258ff43ef9':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:583: undefined reference to `__imp_PyExc_ImportError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `_import_array':\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1525: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1509: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1515: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1537: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1489: undefined reference to `__imp_PyExc_AttributeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1501: undefined reference to `__imp_PyExc_RuntimeError'\\r. collect2.exe: error: ld returned 1 exit status\\r. \", 'FunctionGraph(Elemwise{add,no_inplace}(TensorConstant{12.0}, TensorConstant{12.0}))')",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mException\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\Sean\\OneDrive\\Engineering\\Python\\AdvancedAlgorithmicTrading\\Bayesian Statistics\\4-MarkovChains.ipynb Cell 1\u001b[0m line \u001b[0;36m2\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m basic_model \u001b[39m=\u001b[39m pymc3\u001b[39m.\u001b[39mModel()\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m \u001b[39mwith\u001b[39;00m basic_model:\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m     \u001b[39m# Define our prior belief about the fairness\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m     \u001b[39m# of the coin using a Beta distribution\u001b[39;00m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m     theta \u001b[39m=\u001b[39m pymc3\u001b[39m.\u001b[39;49mBeta(\u001b[39m\"\u001b[39;49m\u001b[39mtheta\u001b[39;49m\u001b[39m\"\u001b[39;49m, alpha\u001b[39m=\u001b[39;49malpha, beta\u001b[39m=\u001b[39;49mbeta)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m     y \u001b[39m=\u001b[39m pymc3\u001b[39m.\u001b[39mBinomial(\u001b[39m\"\u001b[39m\u001b[39my\u001b[39m\u001b[39m\"\u001b[39m, n\u001b[39m=\u001b[39mn, p\u001b[39m=\u001b[39mtheta, observed\u001b[39m=\u001b[39mz)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m     \u001b[39m# Carry out the MCMC analysis using the Metropolis algorithm\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/Sean/OneDrive/Engineering/Python/AdvancedAlgorithmicTrading/Bayesian%20Statistics/4-MarkovChains.ipynb#W0sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m     \u001b[39m# Use Maximum A Posteriori (MAP) optimisation as initial value for MCMC\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\pymc3\\distributions\\distribution.py:124\u001b[0m, in \u001b[0;36mDistribution.__new__\u001b[1;34m(cls, name, *args, **kwargs)\u001b[0m\n\u001b[0;32m    122\u001b[0m     dist \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39m\u001b[39m.\u001b[39mdist(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs, shape\u001b[39m=\u001b[39mshape)\n\u001b[0;32m    123\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m--> 124\u001b[0m     dist \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39;49m\u001b[39m.\u001b[39;49mdist(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[0;32m    125\u001b[0m \u001b[39mreturn\u001b[39;00m model\u001b[39m.\u001b[39mVar(name, dist, data, total_size, dims\u001b[39m=\u001b[39mdims)\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\pymc3\\distributions\\distribution.py:133\u001b[0m, in \u001b[0;36mDistribution.dist\u001b[1;34m(cls, *args, **kwargs)\u001b[0m\n\u001b[0;32m    130\u001b[0m \u001b[39m@classmethod\u001b[39m\n\u001b[0;32m    131\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdist\u001b[39m(\u001b[39mcls\u001b[39m, \u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs):\n\u001b[0;32m    132\u001b[0m     dist \u001b[39m=\u001b[39m \u001b[39mobject\u001b[39m\u001b[39m.\u001b[39m\u001b[39m__new__\u001b[39m(\u001b[39mcls\u001b[39m)\n\u001b[1;32m--> 133\u001b[0m     dist\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(\u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[0;32m    134\u001b[0m     \u001b[39mreturn\u001b[39;00m dist\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\pymc3\\distributions\\continuous.py:1235\u001b[0m, in \u001b[0;36mBeta.__init__\u001b[1;34m(self, alpha, beta, mu, sigma, sd, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1232\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39malpha \u001b[39m=\u001b[39m alpha \u001b[39m=\u001b[39m tt\u001b[39m.\u001b[39mas_tensor_variable(floatX(alpha))\n\u001b[0;32m   1233\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbeta \u001b[39m=\u001b[39m beta \u001b[39m=\u001b[39m tt\u001b[39m.\u001b[39mas_tensor_variable(floatX(beta))\n\u001b[1;32m-> 1235\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmean \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39malpha \u001b[39m/\u001b[39m (\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49malpha \u001b[39m+\u001b[39;49m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mbeta)\n\u001b[0;32m   1236\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mvariance \u001b[39m=\u001b[39m (\n\u001b[0;32m   1237\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39malpha \u001b[39m*\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbeta \u001b[39m/\u001b[39m ((\u001b[39mself\u001b[39m\u001b[39m.\u001b[39malpha \u001b[39m+\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbeta) \u001b[39m*\u001b[39m\u001b[39m*\u001b[39m \u001b[39m2\u001b[39m \u001b[39m*\u001b[39m (\u001b[39mself\u001b[39m\u001b[39m.\u001b[39malpha \u001b[39m+\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mbeta \u001b[39m+\u001b[39m \u001b[39m1\u001b[39m))\n\u001b[0;32m   1238\u001b[0m )\n\u001b[0;32m   1240\u001b[0m assert_negative_support(alpha, \u001b[39m\"\u001b[39m\u001b[39malpha\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mBeta\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\tensor\\var.py:101\u001b[0m, in \u001b[0;36m_tensor_py_operators.__add__\u001b[1;34m(self, other)\u001b[0m\n\u001b[0;32m     99\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__add__\u001b[39m(\u001b[39mself\u001b[39m, other):\n\u001b[0;32m    100\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 101\u001b[0m         \u001b[39mreturn\u001b[39;00m theano\u001b[39m.\u001b[39;49mtensor\u001b[39m.\u001b[39;49mbasic\u001b[39m.\u001b[39;49madd(\u001b[39mself\u001b[39;49m, other)\n\u001b[0;32m    102\u001b[0m     \u001b[39m# We should catch the minimum number of exception here.\u001b[39;00m\n\u001b[0;32m    103\u001b[0m     \u001b[39m# Otherwise this will convert error when Theano flags\u001b[39;00m\n\u001b[0;32m    104\u001b[0m     \u001b[39m# compute_test_value is used\u001b[39;00m\n\u001b[0;32m    105\u001b[0m     \u001b[39m# Evidently, we need to catch NotImplementedError\u001b[39;00m\n\u001b[0;32m    106\u001b[0m     \u001b[39m# TypeError from as_tensor_variable are caught in Elemwise.make_node\u001b[39;00m\n\u001b[0;32m    107\u001b[0m     \u001b[39m# Oterwise TensorVariable * SparseVariable won't work!\u001b[39;00m\n\u001b[0;32m    108\u001b[0m     \u001b[39mexcept\u001b[39;00m (\u001b[39mNotImplementedError\u001b[39;00m, \u001b[39mTypeError\u001b[39;00m):\n\u001b[0;32m    109\u001b[0m         \u001b[39m# We must return NotImplemented and not an\u001b[39;00m\n\u001b[0;32m    110\u001b[0m         \u001b[39m# NotImplementedError or raise an NotImplementedError.\u001b[39;00m\n\u001b[0;32m    111\u001b[0m         \u001b[39m# That way python will give a good error message like this\u001b[39;00m\n\u001b[0;32m    112\u001b[0m         \u001b[39m# `TypeError: unsupported operand type(s) for +:\u001b[39;00m\n\u001b[0;32m    113\u001b[0m         \u001b[39m# 'TensorVariable' and 'TensorVariable'`\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\graph\\op.py:253\u001b[0m, in \u001b[0;36mOp.__call__\u001b[1;34m(self, *inputs, **kwargs)\u001b[0m\n\u001b[0;32m    250\u001b[0m node \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmake_node(\u001b[39m*\u001b[39minputs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n\u001b[0;32m    252\u001b[0m \u001b[39mif\u001b[39;00m config\u001b[39m.\u001b[39mcompute_test_value \u001b[39m!=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39moff\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m--> 253\u001b[0m     compute_test_value(node)\n\u001b[0;32m    255\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdefault_output \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m    256\u001b[0m     rval \u001b[39m=\u001b[39m node\u001b[39m.\u001b[39moutputs[\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdefault_output]\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\graph\\op.py:126\u001b[0m, in \u001b[0;36mcompute_test_value\u001b[1;34m(node)\u001b[0m\n\u001b[0;32m    123\u001b[0m     compute_map[o] \u001b[39m=\u001b[39m [\u001b[39mFalse\u001b[39;00m]\n\u001b[0;32m    125\u001b[0m \u001b[39m# Create a thunk that performs the computation\u001b[39;00m\n\u001b[1;32m--> 126\u001b[0m thunk \u001b[39m=\u001b[39m node\u001b[39m.\u001b[39;49mop\u001b[39m.\u001b[39;49mmake_thunk(node, storage_map, compute_map, no_recycling\u001b[39m=\u001b[39;49m[])\n\u001b[0;32m    127\u001b[0m thunk\u001b[39m.\u001b[39minputs \u001b[39m=\u001b[39m [storage_map[v] \u001b[39mfor\u001b[39;00m v \u001b[39min\u001b[39;00m node\u001b[39m.\u001b[39minputs]\n\u001b[0;32m    128\u001b[0m thunk\u001b[39m.\u001b[39moutputs \u001b[39m=\u001b[39m [storage_map[v] \u001b[39mfor\u001b[39;00m v \u001b[39min\u001b[39;00m node\u001b[39m.\u001b[39moutputs]\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\graph\\op.py:634\u001b[0m, in \u001b[0;36mCOp.make_thunk\u001b[1;34m(self, node, storage_map, compute_map, no_recycling, impl)\u001b[0m\n\u001b[0;32m    630\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mprepare_node(\n\u001b[0;32m    631\u001b[0m     node, storage_map\u001b[39m=\u001b[39mstorage_map, compute_map\u001b[39m=\u001b[39mcompute_map, impl\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mc\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m    632\u001b[0m )\n\u001b[0;32m    633\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 634\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmake_c_thunk(node, storage_map, compute_map, no_recycling)\n\u001b[0;32m    635\u001b[0m \u001b[39mexcept\u001b[39;00m (\u001b[39mNotImplementedError\u001b[39;00m, MethodNotDefined):\n\u001b[0;32m    636\u001b[0m     \u001b[39m# We requested the c code, so don't catch the error.\u001b[39;00m\n\u001b[0;32m    637\u001b[0m     \u001b[39mif\u001b[39;00m impl \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mc\u001b[39m\u001b[39m\"\u001b[39m:\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\graph\\op.py:600\u001b[0m, in \u001b[0;36mCOp.make_c_thunk\u001b[1;34m(self, node, storage_map, compute_map, no_recycling)\u001b[0m\n\u001b[0;32m    598\u001b[0m         \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mDisabling C code for \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mself\u001b[39m\u001b[39m}\u001b[39;00m\u001b[39m due to unsupported float16\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    599\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mNotImplementedError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mfloat16\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m--> 600\u001b[0m outputs \u001b[39m=\u001b[39m cl\u001b[39m.\u001b[39;49mmake_thunk(\n\u001b[0;32m    601\u001b[0m     input_storage\u001b[39m=\u001b[39;49mnode_input_storage, output_storage\u001b[39m=\u001b[39;49mnode_output_storage\n\u001b[0;32m    602\u001b[0m )\n\u001b[0;32m    603\u001b[0m thunk, node_input_filters, node_output_filters \u001b[39m=\u001b[39m outputs\n\u001b[0;32m    605\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mrval\u001b[39m():\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\link\\c\\basic.py:1203\u001b[0m, in \u001b[0;36mCLinker.make_thunk\u001b[1;34m(self, input_storage, output_storage, storage_map)\u001b[0m\n\u001b[0;32m   1175\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m   1176\u001b[0m \u001b[39mCompiles this linker's fgraph and returns a function to perform the\u001b[39;00m\n\u001b[0;32m   1177\u001b[0m \u001b[39mcomputations, as well as lists of storage cells for both the inputs\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1200\u001b[0m \u001b[39m  first_output = ostor[0].data\u001b[39;00m\n\u001b[0;32m   1201\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m   1202\u001b[0m init_tasks, tasks \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mget_init_tasks()\n\u001b[1;32m-> 1203\u001b[0m cthunk, module, in_storage, out_storage, error_storage \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m__compile__(\n\u001b[0;32m   1204\u001b[0m     input_storage, output_storage, storage_map\n\u001b[0;32m   1205\u001b[0m )\n\u001b[0;32m   1207\u001b[0m res \u001b[39m=\u001b[39m _CThunk(cthunk, init_tasks, tasks, error_storage, module)\n\u001b[0;32m   1208\u001b[0m res\u001b[39m.\u001b[39mnodes \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnode_order\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\link\\c\\basic.py:1138\u001b[0m, in \u001b[0;36mCLinker.__compile__\u001b[1;34m(self, input_storage, output_storage, storage_map)\u001b[0m\n\u001b[0;32m   1136\u001b[0m input_storage \u001b[39m=\u001b[39m \u001b[39mtuple\u001b[39m(input_storage)\n\u001b[0;32m   1137\u001b[0m output_storage \u001b[39m=\u001b[39m \u001b[39mtuple\u001b[39m(output_storage)\n\u001b[1;32m-> 1138\u001b[0m thunk, module \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcthunk_factory(\n\u001b[0;32m   1139\u001b[0m     error_storage,\n\u001b[0;32m   1140\u001b[0m     input_storage,\n\u001b[0;32m   1141\u001b[0m     output_storage,\n\u001b[0;32m   1142\u001b[0m     storage_map,\n\u001b[0;32m   1143\u001b[0m )\n\u001b[0;32m   1144\u001b[0m \u001b[39mreturn\u001b[39;00m (\n\u001b[0;32m   1145\u001b[0m     thunk,\n\u001b[0;32m   1146\u001b[0m     module,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1155\u001b[0m     error_storage,\n\u001b[0;32m   1156\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\link\\c\\basic.py:1634\u001b[0m, in \u001b[0;36mCLinker.cthunk_factory\u001b[1;34m(self, error_storage, in_storage, out_storage, storage_map)\u001b[0m\n\u001b[0;32m   1632\u001b[0m     \u001b[39mfor\u001b[39;00m node \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mnode_order:\n\u001b[0;32m   1633\u001b[0m         node\u001b[39m.\u001b[39mop\u001b[39m.\u001b[39mprepare_node(node, storage_map, \u001b[39mNone\u001b[39;00m, \u001b[39m\"\u001b[39m\u001b[39mc\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m-> 1634\u001b[0m     module \u001b[39m=\u001b[39m get_module_cache()\u001b[39m.\u001b[39;49mmodule_from_key(key\u001b[39m=\u001b[39;49mkey, lnk\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m)\n\u001b[0;32m   1636\u001b[0m \u001b[39mvars\u001b[39m \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39minputs \u001b[39m+\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39moutputs \u001b[39m+\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39morphans\n\u001b[0;32m   1637\u001b[0m \u001b[39m# List of indices that should be ignored when passing the arguments\u001b[39;00m\n\u001b[0;32m   1638\u001b[0m \u001b[39m# (basically, everything that the previous call to uniq eliminated)\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\link\\c\\cmodule.py:1191\u001b[0m, in \u001b[0;36mModuleCache.module_from_key\u001b[1;34m(self, key, lnk)\u001b[0m\n\u001b[0;32m   1189\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m   1190\u001b[0m     location \u001b[39m=\u001b[39m dlimport_workdir(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdirname)\n\u001b[1;32m-> 1191\u001b[0m     module \u001b[39m=\u001b[39m lnk\u001b[39m.\u001b[39;49mcompile_cmodule(location)\n\u001b[0;32m   1192\u001b[0m     name \u001b[39m=\u001b[39m module\u001b[39m.\u001b[39m\u001b[39m__file__\u001b[39m\n\u001b[0;32m   1193\u001b[0m     \u001b[39massert\u001b[39;00m name\u001b[39m.\u001b[39mstartswith(location)\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\link\\c\\basic.py:1543\u001b[0m, in \u001b[0;36mCLinker.compile_cmodule\u001b[1;34m(self, location)\u001b[0m\n\u001b[0;32m   1541\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m   1542\u001b[0m     _logger\u001b[39m.\u001b[39mdebug(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mLOCATION \u001b[39m\u001b[39m{\u001b[39;00mlocation\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[1;32m-> 1543\u001b[0m     module \u001b[39m=\u001b[39m c_compiler\u001b[39m.\u001b[39;49mcompile_str(\n\u001b[0;32m   1544\u001b[0m         module_name\u001b[39m=\u001b[39;49mmod\u001b[39m.\u001b[39;49mcode_hash,\n\u001b[0;32m   1545\u001b[0m         src_code\u001b[39m=\u001b[39;49msrc_code,\n\u001b[0;32m   1546\u001b[0m         location\u001b[39m=\u001b[39;49mlocation,\n\u001b[0;32m   1547\u001b[0m         include_dirs\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mheader_dirs(),\n\u001b[0;32m   1548\u001b[0m         lib_dirs\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mlib_dirs(),\n\u001b[0;32m   1549\u001b[0m         libs\u001b[39m=\u001b[39;49mlibs,\n\u001b[0;32m   1550\u001b[0m         preargs\u001b[39m=\u001b[39;49mpreargs,\n\u001b[0;32m   1551\u001b[0m     )\n\u001b[0;32m   1552\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m   1553\u001b[0m     e\u001b[39m.\u001b[39margs \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m (\u001b[39mstr\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfgraph),)\n",
      "File \u001b[1;32mc:\\Users\\Sean\\anaconda3\\envs\\pymc3_env\\lib\\site-packages\\theano\\link\\c\\cmodule.py:2546\u001b[0m, in \u001b[0;36mGCC_compiler.compile_str\u001b[1;34m(module_name, src_code, location, include_dirs, lib_dirs, libs, preargs, py_module, hide_symbols)\u001b[0m\n\u001b[0;32m   2542\u001b[0m     \u001b[39m# We replace '\\n' by '. ' in the error message because when Python\u001b[39;00m\n\u001b[0;32m   2543\u001b[0m     \u001b[39m# prints the exception, having '\\n' in the text makes it more\u001b[39;00m\n\u001b[0;32m   2544\u001b[0m     \u001b[39m# difficult to read.\u001b[39;00m\n\u001b[0;32m   2545\u001b[0m     compile_stderr \u001b[39m=\u001b[39m compile_stderr\u001b[39m.\u001b[39mreplace(\u001b[39m\"\u001b[39m\u001b[39m\\n\u001b[39;00m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39m. \u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m-> 2546\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mException\u001b[39;00m(\n\u001b[0;32m   2547\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCompilation failed (return status=\u001b[39m\u001b[39m{\u001b[39;00mstatus\u001b[39m}\u001b[39;00m\u001b[39m): \u001b[39m\u001b[39m{\u001b[39;00mcompile_stderr\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[0;32m   2548\u001b[0m     )\n\u001b[0;32m   2549\u001b[0m \u001b[39melif\u001b[39;00m config\u001b[39m.\u001b[39mcmodule__compilation_warning \u001b[39mand\u001b[39;00m compile_stderr:\n\u001b[0;32m   2550\u001b[0m     \u001b[39m# Print errors just below the command line.\u001b[39;00m\n\u001b[0;32m   2551\u001b[0m     \u001b[39mprint\u001b[39m(compile_stderr)\n",
      "\u001b[1;31mException\u001b[0m: (\"Compilation failed (return status=1): C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `run':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:99: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:124: undefined reference to `__imp_PyExc_ValueError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:130: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:158: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:195: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:210: undefined reference to `__imp_PyExc_ValueError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:479: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `_Py_INCREF':\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/include/object.h:459: undefined reference to `__imp__Py_NoneStruct'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `run':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:485: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:244: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:265: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:173: undefined reference to `__imp_PyExc_TypeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:179: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:515: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:514: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:516: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:259: undefined reference to `__imp_PyExc_TypeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:279: undefined reference to `__imp__Py_NoneStruct'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:294: undefined reference to `__imp_PyExc_ValueError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:328: undefined reference to `__imp_PyExc_NotImplementedError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:349: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:216: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:343: undefined reference to `__imp_PyExc_TypeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:300: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:397: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `instantiate':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:550: undefined reference to `__imp_PyExc_TypeError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `_import_array':\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1493: undefined reference to `__imp_PyCapsule_Type'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1494: undefined reference to `__imp_PyExc_RuntimeError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `PyInit_mf6917bb35eaa79d4a20c20b9dc13c0435e656b1bdb67265fed3d06258ff43ef9':\\r. C:/Users/Sean/AppData/Local/Theano/compiledir_Windows-10-10.0.19045-SP0-AMD64_Family_23_Model_113_Stepping_0_AuthenticAMD-3.8.18-64/tmpf4mhovs3/mod.cpp:583: undefined reference to `__imp_PyExc_ImportError'\\r. C:\\\\Users\\\\Sean\\\\AppData\\\\Local\\\\Temp\\\\ccdko0oI.o: In function `_import_array':\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1525: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1509: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1515: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1537: undefined reference to `__imp_PyExc_RuntimeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1489: undefined reference to `__imp_PyExc_AttributeError'\\r. c:/Users/Sean/anaconda3/envs/pymc3_env/lib/site-packages/numpy/core/include/numpy/__multiarray_api.h:1501: undefined reference to `__imp_PyExc_RuntimeError'\\r. collect2.exe: error: ld returned 1 exit status\\r. \", 'FunctionGraph(Elemwise{add,no_inplace}(TensorConstant{12.0}, TensorConstant{12.0}))')"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pymc3\n",
    "import scipy.stats as stats\n",
    "\n",
    "plt.style.use(\"ggplot\")\n",
    "n = 50\n",
    "z = 10\n",
    "alpha = 12\n",
    "beta = 12\n",
    "\n",
    "alpha_post = 22\n",
    "beta_post = 52\n",
    "\n",
    "iterations = 100000\n",
    "basic_model = pymc3.Model()\n",
    "\n",
    "with basic_model:\n",
    "    # Define our prior belief about the fairness\n",
    "    # of the coin using a Beta distribution\n",
    "    theta = pymc3.Beta(\"theta\", alpha=alpha, beta=beta)\n",
    "    y = pymc3.Binomial(\"y\", n=n, p=theta, observed=z)\n",
    "    # Carry out the MCMC analysis using the Metropolis algorithm\n",
    "    # Use Maximum A Posteriori (MAP) optimisation as initial value for MCMC\n",
    "    start = pymc3.find_MAP()\n",
    "    # Use the Metropolis algorithm (as opposed to NUTS or HMC, etc.)\n",
    "    step = pymc3.Metropolis()\n",
    "    trace = pymc3.sample(\n",
    "    iterations, step, start, random_seed=1, progressbar=True\n",
    "    )\n",
    "bins=50\n",
    "plt.hist(\n",
    "trace[\"theta\"], bins,density=True,\n",
    "histtype=\"step\",\n",
    "label=\"Posterior (MCMC)\", color=\"red\"\n",
    ")\n",
    "\n",
    "x = np.linspace(0, 1, 100)\n",
    "plt.plot(\n",
    "x, stats.beta.pdf(x, alpha, beta),\n",
    "\"--\", label=\"Prior\", color=\"blue\"\n",
    ")\n",
    "\n",
    "plt.plot(\n",
    "x, stats.beta.pdf(x, alpha_post, beta_post),\n",
    "    label='Posterior (Analytic)', color=\"green\"\n",
    ")\n",
    "\n",
    "plt.legend(title=\"Parameters\", loc=\"best\")\n",
    "plt.xlabel(\"$\\\\theta$, Fairness\")\n",
    "plt.ylabel(\"Density\")\n",
    "plt.show()\n",
    "\n",
    "pymc3.traceplot(trace)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
